### PR TITLE
Switch to stretchr/testify / mockery for mocks

### PR DESCRIPTION
**Pull request by skitt**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

testify is used throughout the codebase; this switches mocks from gomock to testify with the help of mockery for code generation. This removes a dependency from the overall codebase (gomock) and adds one to the tool dependencies (mockery); mockery-generated mocks only depend on testify/mock, not mockery itself.

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

### CODE REVIEW

**Comment by skitt** [comment]
```patch
 --- /logs/artifacts/stats-base.txt	2024-06-19 14:58:00.337670693 +0000
+++ /logs/artifacts/stats.txt	2024-06-19 14:57:55.048677241 +0000
@@ -1,6 +1,6 @@
 Direct Dependencies: 187 
 Transitive Dependencies: 327 
-Total Dependencies: 368 
+Total Dependencies: 367 
 Max Depth Of Dependencies: 24 
 All dependencies:
 
@@ -327,7 +327,6 @@
 go.starlark.net
 go.uber.org/atomic
 go.uber.org/goleak
-go.uber.org/mock
 go.uber.org/multierr
 go.uber.org/zap
 golang.org/x/crypto
```

**Review by sanchayanghosh — COMMENTED** [review]
I see some dependencies have been updated in the code. Can we have a seperate PR for version changes? 

Basically this PR only modifies the existing go-mock with testify

And further PRs update version changes

**Comment by dims** [comment]
> I see some dependencies have been updated in the code. Can we have a seperate PR for version changes?

this is ok. this is a complicated thing we are trying to unravel

**Comment by dims** [comment]
@skitt looks like a LOT of the changes were generated code. good to see the update/verify mock pattern still applies and we don't have to do anything new. thanks for keeping the root go mod file clean as well. overall it looks good. let's get some eyes from different sigs whose test cases we touched.

File: hack/update-mocks.sh [hunk 1]
```diff
@@ -42,17 +42,22 @@ function git_grep() {
...
-GENERATED_MOCK_FILE_REGEX="^// Code generated by MockGen. DO NOT EDIT.$"
+GENERATED_MOCK_FILE_REGEX="^// Code generated by mockery v[0-9.]\+. DO NOT EDIT.$"
 
 # pick a tempfile path for writing to
 tmp=$(mktemp)
 kube::util::trap_add "rm -f ${tmp:?}" EXIT
 
+# test/utils/oidc/testserver.go needs the mocks to build; mockery processes full packages,
+# and fails if testserver.go can't be built. Rename it out of the way for mock generation.
+kube::util::trap_add "mv test/utils/oidc/testserver.go.disabled test/utils/oidc/testserver.go" EXIT
```
Comment (liggitt) [inline hack/update-mocks.sh:53]: hrmm... @enj @aramase is there a way to tweak this test so this juggling isn't needed?

Comment (skitt) [reply hack/update-mocks.sh:53]: Moving `testserver.go` or the generated mocks to another package would fix this.

Comment (liggitt) [reply hack/update-mocks.sh:53]: can we move the generated mocks to test/utils/oidc/mock/... as part of this PR? we're already regenerating completely new files, right?

Comment (skitt) [reply hack/update-mocks.sh:53]: Yes, we are indeed. The mocks are only used in `testserver.go` so the change wouldn’t affect anyone else. While I’m at it I’ll fix the ugly `mock_jw_ks_handler.go` file name...

**Review by BenTheElder — COMMENTED** [review]
/lgtm
/approve

File: hack/update-mocks.sh (hunk 1 above)
Comment (BenTheElder) [reply hack/update-mocks.sh]: this seems better, thanks

**Pull request merged**